test/run.py

```diff
@@ -0,0 +1,248 @@
+    parser.add_argument(
+        'test_dirs', nargs='*', type=str, default=None)
+    parser.add_argument('--controller-tests-parameters', type=str, nargs='+',
+                        default=[])
+    parser.add_argument('--bless', action='store_true', default=False)
+    parser.add_argument('--in-parallel', action='store_true', default=False)
+    parser.add_argument('--end-to-end-only', action='store_true', default=False)
+    parser.add_argument('--controller-tests-only', action='store_true',
+                        default=False)
+    parser.add_argument('--clean-up-first', action='store_true', default=False)
+
+    options = parser.parse_args(sys.argv[1:])
+
+    if options.end_to_end_only and options.controller_tests_only:
+        raise ValueError('Options end_to_end_only and controller_tests_only '
+                         'can\'t be specified simultaneously')
+    end_to_end_only = options.end_to_end_only \
+        or len(options.test_dirs or []) > 0
+
+    end_to_end_tests = end_to_end_only or not options.controller_tests_only
+    controller_tests = options.controller_tests_only or not end_to_end_only
+
+    if options.clean_up_first:
+        test_utils.cleanup(interrupted=False)
+
+    signal.signal(signal.SIGTERM, test_utils.sig_cleanup)
+    with test_utils.DoCleanupContextManager() as cleanup_manager:
+        try:
+            passed = True
+            os.makedirs(DATA_DIRECTORY, exist_ok=True)
+            if end_to_end_tests:
+                build_cli()
+
+            if 'PLZ_HOST' in os.environ:
+                plz_host = os.environ['PLZ_HOST']
+                if 'PLZ_PORT' not in os.environ:
+                    raise ValueError('PLZ_HOST is defined but PLZ_PORT is not!')
+                plz_port = int(os.environ['PLZ_PORT'])
+                network = 'host'
+            else:
+                start_controller()
+                plz_host = CONTROLLER_HOSTNAME
+                plz_port = CONTROLLER_PORT
+                network = get_network()
+
+            if controller_tests:
+                if not run_controller_tests(
+                        network, plz_host, plz_port,
+                        options.controller_tests_parameters):
+                    test_utils.print_error('Some controller tests failed!')
+                    passed = False
+
+            if end_to_end_tests:
+                failed_tests = run_end_to_end_tests(
+                    network, plz_host, plz_port, options.bless,
+                    options.in_parallel,
+                    options.test_dirs)
+                if len(failed_tests) != 0:
+                    passed = False
+                    test_utils.print_error('Some end-to-end tests failed!')
+                for t in failed_tests:
+                    test_utils.print_error(t)
+        except InterruptedError:
+            cleanup_manager.interrupted = True
+            raise
+
+    return 1 if not passed else 0
+
+
+if __name__ == '__main__':
+    sys.exit(main())
```

Comment:
This seems like a good opportunity to introduce a standard formatter such as [Black](https://black.readthedocs.io/).